### ISSUE TITLE
Rule to prevent assigning any value to a variable named undefined

### ISSUE THREAD

**Issue by rightaway** [eslint/eslint#8031]
The docs for the `no-undefined` rule gives this example. I don't want to use the `no-undefined` rule, since I use undefined in many places. But it would be useful to have a rule that doesn't allow the shadowing of undefined to happen, where you can't assign a value to a variable named `undefined` in any way. (Like var, let, const, destructuring, function arguments, etc)

```js
function doSomething(data) {
    var undefined = "hi";

    // doesn't do what you think it does
    if (data === undefined) {
        // ...
    }

}
```

**Comment by platinumazure** [comment]
Not sure I understand. You could use `no-undefined` but then use `void 0` when you need to specify undefined as a value, wouldn't that work?

**Comment by platinumazure** [comment]
Otherwise, I could see a good case for a `no-restricted-declarations` rule which would forbid identifiers in declarations according to config (analogous to `no-restricted-globals`). Perhaps a similar case could be made for restricting assignments (even without a declaration) as well.

**Comment by nzakas** [comment]
This is pretty specific, so I'd suggest creating a custom rule instead.

**Comment by rightaway** [comment]
> You could use no-undefined but then use void 0 when you need to specify undefined as a value, wouldn't that work?

I think it would still be good to be able to use undefined instead of `void 0`, which other people reading your code might not be familiar with.

**Comment by platinumazure** [comment]
Fair enough. In that case, though, I agree with @nzakas and don't think
this is something we can change in ESLint core.

**Comment by not-an-aardvark** [comment]
It looks like there is consensus that this would be a better fit for a custom rule, so I'm going to close this issue. Thanks anyway for the suggestion -- if you do end up implementing a custom rule for this, feel free to link it here.
